`ShellJob`: Change the signature of custom parser functions (#97)
sphuber/aiida-shell: docs/source/howto.rst

    which prints ``{'a': 1}``.

Optionally, the parsing function can also define the ``parser`` argument.
If defined, in addition to the ``dirpath``, the parser receives an instance of the ``Parser`` class.
This instance can be useful for a number of things, such as:

* Access the logger in order to log messages

> * Access the node the represents the job, through which input nodes can be accessed

Reviewer: ```suggestion
* Access the node that represents the `ShellJob`, from which e.g. its input nodes can be accessed
```

Author: Done

Below is an example of how the ``parser`` argument can be put to use:

.. code-block:: python

    from pathlib import Path
    from aiida_shell import launch_shell_job
    from aiida.parsers import Parser

    def custom_parser(dirpath: Path, parser: Parser):
        from aiida.orm import Bool, Str

        inputs = parser.node.inputs  # Retrieve inputs of the job

        if inputs.arguments[0] == 'return-bool':
            parser.logger.warning('Arguments set to `return-bool`, returning a bool')
            return {'output': Bool(True)}
        else:
            return {'output': Str((dirpath / 'stdout').read_text().strip())}

    results, node = launch_shell_job(
        'echo',
        arguments='return-bool',
        parser=custom_parser
    )
    print(results['output'].value)

which should print

.. code-block:: console

    07/18/2024 03:49:32 PM <13555> aiida.parser.ShellParser: [WARNING] Arguments set to `return-bool`, returning a bool
    True

> 

Reviewer: Maybe we can add this for clarification:
```rst
.. note:
As shown in the code snippet above, you don't need to instantiate a `Parser` yourself. The `ShellParser` associated with the `ShellJob` that is being executed will be assigned automatically, and the functionality you provide in your `custom_parser` function will be made available to it. This is why you can access the attributes of the `CalcJobNode`, e.g. its input arguments, in your parsing function.
```

Author: I am not sure this would be helping the "common" user much but might be more confusing. The whole idea is that they won't be really familiar and don't have to be with parser plugins. So So it should be obvious tht they don't have to instantiate it themselves, because they never do and don't even really know what it is. 

I was thinking to add a topics section at some point that essentially explains a bit how `aiida-shell` works under the hood for users of AiiDA that are familiar with `CalcJob` and `Parser` plugins, so they can more easily integrate and tweak it themselves in their workflows. This would be a perfect place for the clarification that you suggested, so I might add it there.

Reviewer: > I am not sure this would be helping the "common" user much but might be more confusing. The whole idea is that they won't be really familiar and don't have to be with parser plugins. So So it should be obvious tht they don't have to instantiate it themselves, because they never do and don't even really know what it is.

Fair enough. The point of confusion for me, initially, was that I define this argument, but then don't pass anything to it (though, this is also the case for `dirpath`). It makes sense now, as I understood that one just passes the function as a `Callable`, but a beginner Python user might not be familiar with this concept. In any case, I think it is clear with the provided example how it should be used.

> I was thinking to add a topics section at some point that essentially explains a bit how `aiida-shell` works under the hood for users of AiiDA that are familiar with `CalcJob` and `Parser` plugins, so they can more easily integrate and tweak it themselves in their workflows. This would be a perfect place for the clarification that you suggested, so I might add it there.

I agree, that would be a good place to explain then how it works.
